compiler/rustc_abi/src/layout.rs

```diff
@@ -897,42 +897,58 @@ pub trait LayoutCalculator {
         }
 
-        let optimize = !repr.inhibit_union_abi_opt();
+        let mut optimize = !repr.inhibit_union_abi_opt();
         let mut size = Size::ZERO;
-        let mut abi = Abi::Aggregate { sized: true };
+        let mut common_non_zst_abi_and_align: Option<(Abi, AbiAndPrefAlign)> = None;
         let index = V::new(0);
         for field in &variants[index] {
             assert!(field.is_sized());
-            align = align.max(field.align);
 
-            // If all non-ZST fields have the same ABI, forward this ABI
-            if optimize && !field.is_zst() {
+            if !field.is_zst() && optimize {
                 // Discard valid range information and allow undef
-                let field_abi = match field.abi {
-                    Abi::Scalar(x) => Abi::Scalar(x.to_union()),
-                    Abi::ScalarPair(x, y) => Abi::ScalarPair(x.to_union(), y.to_union()),
-                    Abi::Vector { element: x, count } => {
-                        Abi::Vector { element: x.to_union(), count }
+                let field_abi = field.abi.to_union();
+
+                if let Some((abi, align)) = &mut common_non_zst_abi_and_align {
+                    if *abi != field_abi {
+                        // Different fields have different ABI: disable opt
+                        optimize = false;
+                    } else {
+                        // Fields with the same non-Aggregate ABI should also
+                        // have the same alignment
+                        if !matches!(abi, Abi::Aggregate { .. }) {
+                            assert_eq!(
+                                align.abi, field.align.abi,
+                                "non-Aggregate field with matching ABI but differing alignment"
+                            );
+                        }
                     }
-                    Abi::Uninhabited | Abi::Aggregate { .. } => Abi::Aggregate { sized: true },
-                };
-
-                if size == Size::ZERO {
-                    // first non ZST: initialize 'abi'
-                    abi = field_abi;
-                } else if abi != field_abi {
-                    // different fields have different ABI: reset to Aggregate
-                    abi = Abi::Aggregate { sized: true };
+                } else {
+                    common_non_zst_abi_and_align = Some((field_abi, field.align));
                 }
             }
 
+            align = align.max(field.align);
             size = cmp::max(size, field.size);
         }
 
         if let Some(pack) = repr.pack {
             align = align.min(AbiAndPrefAlign::new(pack));
         }
 
+        // If all non-ZST fields have the same ABI, we may forward that ABI
+        // for the union as a whole, unless otherwise inhibited.
+        let abi = match (optimize, common_non_zst_abi_and_align) {
+            (false, _) | (_, None) => Abi::Aggregate { sized: true },
+            (true, Some((abi, _))) => {
```

Comment:
You're not using the `align` in `common_non_zst_abi_and_align`, so you can remove that and be left with just `common_non_zst_abi`, I think?

Reply:
Wasn't using it down here but above in the loop for the "non-Aggregate field with matching ABI but differing alignment" check. Renamed things to try to make it a little more clear.